### PR TITLE
Allow UTF-8 names for registered members

### LINKED ISSUE #17: SourceGenerators Property naming does not support utf8
### Godot version

4.3.stable

### Godot .NET packages version

4.3.0-ci(ddbd6bd)

### System information

Arch Linux 6.10.7-zen1-1.1-zen Intel(R) Core(TM) i5-9300H

### .NET information

.NET SDK:  Version:           9.0.100-preview.7.24407.12  Commit:            d672b8a045  Workload version:  9.0.100-manifests.2aef0cee  MSBuild version:   17.12.0-preview-24374-02+48e81c6f1  运行时环境:  OS Name:     arch  OS Version:    OS Platform: Linux  RID:         linux-x64  Base Path:   /usr/share/dotnet/sdk/9.0.100-preview.7.24407.12/  已安装 .NET 工作负载: 配置为在安装新清单时使用 loose manifests。 没有要显示的已安装工作负载。  Host:   Version:      9.0.0-preview.7.24405.7   Architecture: x64   Commit:       static  .NET SDKs installed:   6.0.423 [/usr/share/dotnet/sdk]   7.0.120 [/usr/share/dotnet/sdk]   8.0.401 [/usr/share/dotnet/sdk]   9.0.100-alpha [/usr/share/dotnet/sdk]   9.0.100-preview.7.24407.12 [/usr/share/dotnet/sdk]  .NET runtimes installed:   Microsoft.NETCore.App 6.0.31 [/usr/share/dotnet/shared/Microsoft.NETCore.App]   Microsoft.NETCore.App 7.0.20 [/usr/share/dotnet/shared/Microsoft.NETCore.App]   Microsoft.NETCore.App 8.0.8 [/usr/share/dotnet/shared/Microsoft.NETCore.App]   Microsoft.NETCore.App 9.0.0-alpha.1.24066.33 [/usr/share/dotnet/shared/Microsoft.NETCore.App]   Microsoft.NETCore.App 9.0.0-preview.7.24405.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]  Other architectures found:   None  Environment variables:   DOTNET_ROOT       [/usr/share/dotnet]  global.json file:   Not found  Learn more:   https://aka.ms/dotnet/info  Download .NET:   https://aka.ms/dotnet/download

### Issue description

Code:
```
public enum 境界 {
	武徒,
	炼精化气,
	炼气化神,
	炼神还虚,
	铜皮铁骨,
	毫发不爽,
	心领神会,
	滴血重生,
	合道同归,
	独步乾坤,
	武圣
}

[GodotClass]
public partial class Game : Node2D {
	[BindProperty]
	public 境界 境界 { get; set; }

	override protected void _Ready() { GD.PrintErr(this); }
}
```

Error:
```
Unhandled Exception: System.TypeInitializationException: A type initializer threw an exception. To determine which type, inspect the InnerException's StackTrace property.
 ---> System.ArgumentException: The bytes are not using ASCII encoding or contains invalid characters. (Parameter 'ascii')
   at Godot.StringName.CreateStaticFromAscii(ReadOnlySpan`1) + 0x5c
   at 球武道.Game.PropertyName..cctor() + 0x3e
   at System.Runtime.CompilerServices.ClassConstructorRunner.EnsureClassConstructorRun(StaticClassConstructionContext*) + 0xb4
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.ClassConstructorRunner.EnsureClassConstructorRun(StaticClassConstructionContext*) + 0x147
   at System.Runtime.CompilerServices.ClassConstructorRunner.CheckStaticClassConstructionReturnGCStaticBase(StaticClassConstructionContext*, Object) + 0x9
   at 球武道.Game.BindMethods(ClassDBRegistrationContext context) + 0x138
   at Godot.Bridge.ClassDB.RegisterClassCore[T](Boolean, Boolean, Boolean, Boolean, Action`1) + 0x551
   at 球武道.ClassDBExtensions.InitializeUserTypes(InitializationLevel) + 0x6e
   at Godot.Bridge.GodotBridge.InitializeLevel_Native(Void*, GDExtensionInitializationLevel) + 0x2f
```

Generated:
图片

### Steps to reproduce

As shown above

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by scgm0**
Fix #17

### CODE REVIEW

File: src/Godot.SourceGenerators/BindMethodsWriter.cs [hunk 1]
```diff
@@ -140,7 +140,12 @@ private static void WriteCachedStringNames(IndentedStringBuilder sb, GodotClassS
 
         void AddCachedStringName(string symbolName, string value)
         {
-            sb.AppendLine($$"""public static global::Godot.StringName @{{symbolName}} { get; } = global::Godot.StringName.CreateStaticFromAscii("{{value}}"u8);""");
+            if (value.IsAscii()) {
+                sb.AppendLine($$"""public static global::Godot.StringName @{{symbolName}} { get; } = global::Godot.StringName.CreateStaticFromAscii("{{value}}"u8);""");
+            } else {
+                sb.AppendLine($$"""public static global::Godot.StringName @{{symbolName}} { get; } = global::Godot.StringName.CreateFromUtf8("{{value}}"u8);""");
```
Comment (raulsntos) [inline src/Godot.SourceGenerators/BindMethodsWriter.cs:146]: To match our code-style, we use the "Allman" style which also matches [C# style guidelines](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#style-guidelines).

```suggestion
            if (value.IsAscii())
            {
                sb.AppendLine($$"""public static global::Godot.StringName @{{symbolName}} { get; } = global::Godot.StringName.CreateStaticFromAscii("{{value}}"u8);""");
            }
            else
            {
                sb.AppendLine($$"""public static global::Godot.StringName @{{symbolName}} { get; } = global::Godot.StringName.CreateFromUtf8("{{value}}"u8);""");
```

File: src/Godot.SourceGenerators/BindMethodsWriter.cs [hunk 2]
```diff
@@ -421,7 +426,11 @@ private static void AppendParameterInfo(this IndentedStringBuilder sb, GodotProp
             : parameter.SymbolName;
 
         sb.Append("new global::Godot.Bridge.ParameterInfo(");
-        sb.Append($"""global::Godot.StringName.CreateStaticFromAscii("{nameValue}"u8), """);
+        if (nameValue.IsAscii()) {
+            sb.Append($"""global::Godot.StringName.CreateStaticFromAscii("{nameValue}"u8), """);
+        } else {
+            sb.Append($"""global::Godot.StringName.CreateFromUtf8("{nameValue}"u8), """);
+        }
```
Comment (raulsntos) [inline src/Godot.SourceGenerators/BindMethodsWriter.cs:433]:
```suggestion
        if (nameValue.IsAscii())
        {
            sb.Append($"""global::Godot.StringName.CreateStaticFromAscii("{nameValue}"u8), """);
        }
        else
        {
            sb.Append($"""global::Godot.StringName.CreateFromUtf8("{nameValue}"u8), """);
        }
```

File: src/Godot.SourceGenerators/BindMethodsWriter.cs [hunk 3]
```diff
@@ -501,7 +510,11 @@ private static void AppendPropertyInfoObjectInitializer(IndentedStringBuilder sb
         }
         if (!string.IsNullOrEmpty(marshalInfo.ClassName))
         {
-            lines.Add($"""ClassName = global::Godot.StringName.CreateStaticFromAscii("{marshalInfo.ClassName}"u8),""");
+            if (marshalInfo.ClassName!.IsAscii()) {
+                lines.Add($"""ClassName = global::Godot.StringName.CreateStaticFromAscii("{marshalInfo.ClassName}"u8),""");
+            } else {
+                lines.Add($"""ClassName = global::Godot.StringName.CreateFromUtf8("{marshalInfo.ClassName}"u8),""");
+            }
```
Comment (raulsntos) [inline src/Godot.SourceGenerators/BindMethodsWriter.cs:517]:
```suggestion
            if (marshalInfo.ClassName!.IsAscii())
            {
                lines.Add($"""ClassName = global::Godot.StringName.CreateStaticFromAscii("{marshalInfo.ClassName}"u8),""");
            }
            else
            {
                lines.Add($"""ClassName = global::Godot.StringName.CreateFromUtf8("{marshalInfo.ClassName}"u8),""");
            }
```

**Review by raulsntos — COMMENTED** [review]
Thanks! Looks good. Since this PR also covers signals and methods, could you change the commit message to reflect that? For example, something like `Allow UTF-8 names for registered members`.

**Comment by scgm0** [comment]
> Thanks! Looks good. Since this PR also covers signals and methods, could you change the commit message to reflect that? For example, something like `Allow UTF-8 names for registered members`.

Thank you for the new name, I forgot to use the common c# code format, please wait.

**Comment by scgm0** [comment]
> Thanks! Looks good. Since this PR also covers signals and methods, could you change the commit message to reflect that? For example, something like `Allow UTF-8 names for registered members`.

Done.

**Review by raulsntos — APPROVED** [review]

**Comment by raulsntos** [comment]
Thanks!

**Pull request merged**